src/rewriter.fs

```diff
@@ -285,8 +291,79 @@ let rec iterateSimplifyAndInline li =
     let simplified = mapTopLevel (mapEnv simplifyExpr simplifyStmt) li
     if foundInlinable then iterateSimplifyAndInline simplified else simplified
 
+let inlineAllConsts li =
+    let mapInnerDecl = function
+        // Unconditional inlining of anything marked "const" -- trust that the
+        // compiler would have yelled if it weren't really really const, so we
+        // can brutishly just inline it.
+        | ({typeQ = tyQ}, defs) as d when List.contains "const" tyQ ->
+            List.iter (fun (def:DeclElt) -> def.name.Inline()) defs
+            d
+        | d -> d
+    let mapStmt = function
+        | Decl d -> Decl (mapInnerDecl d)
+        | s -> s
+    let mapExpr _ e = e
+    let mapTLDecl = function
+        | TLDecl d -> TLDecl (mapInnerDecl d)
+        | d -> d
+    li
+    |> mapTopLevel (mapEnv mapExpr mapStmt)
+    |> List.map mapTLDecl
+
+let markLValues li =
+    // Helpers for the bodies of functions: find any expression of the form
+    // "foo = ..." or "foo += ..." etc, then scan through all of "foo" marking
+    // any variable seen there as potentially in an lvalue position. This will
+    // over-mark things, e.g., "x[i]" both "x" and "i" will get marked even
+    // though the latter does not need to be, but it is simple.
+    let markVars env = function
+        | Var v as e ->
+            match env.vars.TryFind v.Name with
+            | Some (_, {name = vv}) -> vv.MarkLValue(); e
+            | _ -> e
+        | e -> e
+    let assignOps = Set.ofList ["="; "+="; "-="; "*="; "/="; "%=";
+        "<<="; ">>="; "&="; "^="; "|="; "_++"; "_--"; "$++"; "$--"]
+    let findWrites env = function
+        | FunCall(Op o, e::args) when Set.contains o assignOps ->
+            let newEnv = {env with fExpr = markVars}
+            FunCall(Op o, (mapExpr newEnv e)::args)
+        | FunCall(Var v, _) as e ->
+            let fnIsMarked =
+                match env.fns.TryFind v.Name with
+                | Some fct when fct.fName.IsLValue -> true
+                | _ -> false
+            if fnIsMarked then
```

Comment:
I don't think you need fnIsMarked:

```f#
match env.fns.TryFind v.Name with
    | Some fct when fct.fName.IsLValue ->
        let newEnv = {env with fExpr = markVars}
        mapExpr newEnv e
    | _ -> e
```

Reply:
Er, yes, of course, not sure what I was thinking here.